DOCS-6057 Profiler guide - Optimizing performance of monolithic services (#19842)
DataDog/documentation: content/en/profiler/guide/isolate-outliers-in-monolithic-services.md

{{< img src="profiler/guide-monolithic-outliers/1-outliers-monolith-cpu-usage.png" alt="" style="width:100%;" >}}

This doesn't provide concrete pointers to the answers, but you can see some anomalous peaks in CPU usage. 

Select the **Show - Avg of** dropdown (highlighted) and change the graph to show `CPU Cores for Top Endpoints` instead. This graph shows how different parts of the application contribute differently to the overall CPU utilization:

{{< img src="profiler/guide-monolithic-outliers/2-outliers-monolith-cpu-top-endpoints.png" alt="" style="width:100%;" >}}

The yellow peaks show that the `GET /store_history` endpoint has some intermittent usage corresponding to the anomalies identified above. However, the peaks might be due to differences in traffic to that endpoint. To understand if profiles can provide further insights, change the metric to `CPU - Average Time Per Call for Top Endpoints`:

{{< img src="profiler/guide-monolithic-outliers/3-outliers-monolith-cpu-avg-time-per-call.png" alt="" style="width:100%;" >}}

> Now you can see that there is an intermittent spike in CPU utilization _for each call_ to `GET /store_history`, rather than the increase being because of more calls.

Reviewer: ```suggestion
Now you can see that there is an intermittent spike in CPU utilization where  _each call_ to `GET /store_history` takes on average 3s of CPU Time, rather than the increase being because of more calls.
```

Author: ```suggestion
Now you can see that there is an intermittent spike in CPU utilization where _each call_ to `GET /store_history` takes on average three seconds of CPU time, rather than the increase being because of more calls.
```

So, the CPU utilization spikes are not due to an increase in traffic but rather an increase in CPU usage per request. 

## Isolate the impact of endpoints

To figure out what is causing increased CPU usage each time `GET /store_history` is called, look at the profiling flame graph for this endpoint during one of the spikes. Select a time range where `GET /store_history` is showing more CPU utilization, scoping the profiling page to that time range. Then switch to the **Flame Graph** visualization to see the methods using the CPU at this time:

{{< img src="profiler/guide-monolithic-outliers/4-outliers-monolith-flame-graph.png" alt="Your image description" style="width:100%;" >}}

> To better understand why the `GET /store_history` endpoint in particular is using more CPU, look at the table on the right-hand side, where you can see the endpoint is second from the top. Select just that row to focus the flame graph specifically on the CPU utilization caused by the `GET /store_history` endpoint. 

Reviewer 2: ```suggestion
To better understand why the `GET /store_history` endpoint is using more CPU, refer to the table highlighted in the previous image, where the endpoint is second from the top. Select that row to focus the flame graph on the CPU utilization caused by the `GET /store_history` endpoint. 
```

> Because you are investigating resource usage per request, also change the value in the drop-down menu at top of the table to `CPU Time per Endpoint Call`, to show the average resource usage per call to that endpoint instead of the average resource usage per minute.

Reviewer 2: ```suggestion
Because you are investigating resource usage per request, also change the value in the dropdown at top of the table to `CPU Time per Endpoint Call`. This shows the average resource usage per call to that endpoint instead of the average resource usage per minute.
```

## Comparing flame graphs

> Now that the graph shows data for the right time and endpoint, you probably have enough data to understand what is causing the spike in CPU utilization. But if not, you can compare the flame graph for the spike with a time when utilization was more acceptable.

Reviewer 2: ```suggestion
With the graph displaying data for the correct time and endpoint, you should have enough data to determine what is causing the spike in CPU utilization. If you're still uncertain, you can compare the flame graph for the spike with a time when utilization was more acceptable.
```

> To see if there are differences in which methods are using a lot of CPU time between a spike and normal usage, click **Compare** (on the far right side of the search field) and select `Previous 15 minutes`. This opens the Comparison view. 

Reviewer 2: ```suggestion
To compare which methods are using a lot of CPU time between a spike and normal usage, click **Compare** (next to the search field) and select `Previous 15 minutes`. This opens the Comparison view. 
```

Author: ```suggestion
To see if there are differences in which methods are using a lot of CPU time between a spike and normal usage, click **Compare** (next to the search field) and select `Previous 15 minutes`. This opens the Comparison view. 
```

> The view shows two graphs, labeled `A` and `B`, each representing a time range for CPU utilization per call for `GET /store_history`. Adjust the time selector for `A` so that it is scoped to a period with low CPU utilization per call:

Reviewer 2: ```suggestion
The view shows two graphs, labeled **A** and **B**, each representing a time range for CPU utilization per `GET /store_history` call. Adjust the time selector for **A** so that it is scoped to a period with low CPU utilization per call:
```

{{< img src="profiler/guide-monolithic-outliers/5-outliers-monolith-compare-flame-graphs.png" alt="Your image description" style="width:100%;" >}}

> The comparison shows the different methods causing CPU utilization on the right side (timeframe `B` during the spike) that are not used at all in timeframe `A` (normal CPU). The culprit, `Product.loadAssets(int)`, is causing the spikes. 

Reviewer: Maybe we should highlight it on the image above. All frames being purple are not helping a lot the readability of the comparison

Reviewer 3: I think ideally we want to hide the Datadog frames to the right of the loadAssets call. They're caused by IAST overhead.

> The comparison shows the different methods causing CPU utilization on the right side (timeframe `B` during the spike) that are not used at all in timeframe `A` (normal CPU). The culprit, `Product.loadAssets(int)`, is causing the spikes. 

Reviewer 2: ```suggestion
The comparison reveals the different methods causing CPU utilization during the spike (timeframe **B**) that are not used during normal CPU usage (timeframe **A**). As shown in the previous image,`Product.loadAssets(int)`, is causing the spikes.
```

> To fix the problem, optimize the method. Looking at the method code, the signature is `Product(int id, String name, boolean shouldLoadAssets)` and we do not need to load assets for the response to the `GET /store_history` endpoint. This implies that there is a bug further up the call stack that improperly instructs the `Product` constructor to load assets.

Author: ```suggestion
To fix the problem, optimize the method. Looking at the method code, the signature is `Product(int id, String name, boolean shouldLoadAssets)` and you do not need to load assets for the response to the `GET /store_history` endpoint. This implies that there is a bug further up the call stack that improperly instructs the `Product` constructor to load assets.
```

Fix that bug and verify that the spikes go away, using the timeseries graphs discussed earlier.

## Isolate the impact of operations (Java)

> There are other attributes available in the profiler. For example, you can filter and group a flame graph by operation names, rather than by functions or threads. For a monolithic application this can be more helpful, as it can more directly state which resources are utilizing more CPU, even if they are shared between endpoints.

Reviewer 2: ```suggestion
There are other attributes available in the profiler. For example, you can filter and group a flame graph by operation names, rather than by functions or threads. For monolithic applications, this can more clearly identify CPU-intensive resources, even if they are shared between endpoints.
```

> The APM `Trace operation` attribute lets you filter and group a flame graph with the same granularity as what's in the traces for the selected endpoints. This is a good middle between the high granularity of threads or methods, and the low granularity of entire endpoints. To isolate operations, select `Trace Operation` from the **CPU time by** drop-down: 

Reviewer 2: ```suggestion
The APM `Trace operation` attribute lets you filter and group a flame graph with the same granularity as the traces for the selected endpoints. This is a good balance between the high granularity of threads or methods, and the low granularity of entire endpoints. To isolate operations, select `Trace Operation` from the **CPU time by** dropdown: 
```

{{< img src="profiler/guide-monolithic-outliers/7-outliers-monolith-trace-operation.png" alt="Your image description" style="width:100%;" >}}

> Now you can see that the `ModelTraining` operation is taking more CPU time than its primary use in the `GET /train` endpoint, so it must be used in other places. Select it to find out where else it is used. In this case, `ModelTraining` is also use by `POST /update_model`.

Reviewer: ```suggestion
Now you can see that the `ModelTraining` operation is taking more CPU time than its primary use in the `GET /train` endpoint, so it must be used in other places. Click on it to find out where else it is used. In this case, `ModelTraining` is also use by `POST /update_model`.
```

We had the feedback that the summary table is not fully intuitive, let's clarify what the exact action is

> ## Isolate your own business logic

Reviewer: ```suggestion
## Isolate your own business logic (Java)
```

> Now you can see that the `ModelTraining` operation is taking more CPU time than its primary use in the `GET /train` endpoint, so it must be used in other places. Click on it to find out where else it is used. In this case, `ModelTraining` is also use by `POST /update_model`.

Reviewer 2: ```suggestion
In the previous image, notice that the `ModelTraining` operation is taking more CPU time than its primary use in the `GET /train` endpoint, so it must be used elsewhere. Click the operation name to determine where else it is used. In this case, `ModelTraining` is also use by `POST /update_model`.
```

> Endpoint and operation isolation is available in your profiles by default, but you may want to isolate a different piece of logic. For example, if the monolith is sensitive to specific customers, you can add a custom filtering to the profiles:

Reviewer 2: ```suggestion
Endpoint and operation isolation is available in your profiles by default, but you may want to isolate a different piece of logic. For example, if the monolith is sensitive to specific customers, you can add a custom filter to the profiles:
```
